libs/template/schema.go

```diff
@@ -0,0 +1,145 @@
+package template
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"reflect"
+)
+
+// This is a JSON Schema compliant struct that we use to do validation checks on
+// the provided configuration
+type Schema struct {
+	// A list of properties that can be used in the config
+	Properties map[string]Property `json:"properties"`
+}
+
+type PropertyType string
+
+const (
+	PropertyTypeString  = PropertyType("string")
+	PropertyTypeInt     = PropertyType("integer")
+	PropertyTypeNumber  = PropertyType("number")
+	PropertyTypeBoolean = PropertyType("boolean")
+)
+
+type Property struct {
+	Type        PropertyType `json:"type"`
+	Description string       `json:"description"`
+	Default     any          `json:"default"`
+}
+
+// function to check whether a float value represents an integer
+func isIntegerValue(v float64) bool {
+	return v == float64(int(v))
+}
+
+// cast value to integer for config values that are floats but are supposed to be
+// integers according to the schema
+//
+// Needed because the default json unmarshaler for maps converts all numbers to floats
+func castFloatConfigValuesToInt(config map[string]any, schema *Schema) error {
+	for k, v := range config {
+		// error because all config keys should be defined in schema too
+		if _, ok := schema.Properties[k]; !ok {
+			return fmt.Errorf("%s is not defined as an input parameter for the template", k)
+		}
+
+		// skip non integer fields
+		fieldInfo := schema.Properties[k]
+		if fieldInfo.Type != PropertyTypeInt {
+			continue
+		}
+
+		// convert floating point type values to integer
+		valueType := reflect.TypeOf(v)
+		switch valueType.Kind() {
+		case reflect.Float32:
+			floatVal := v.(float32)
+			if !isIntegerValue(float64(floatVal)) {
+				return fmt.Errorf("expected %s to have integer value but it is %v", k, v)
+			}
+			config[k] = int(floatVal)
+		case reflect.Float64:
+			floatVal := v.(float64)
+			if !isIntegerValue(floatVal) {
+				return fmt.Errorf("expected %s to have integer value but it is %v", k, v)
+			}
+			config[k] = int(floatVal)
+		}
+	}
+	return nil
+}
+
+func assignDefaultConfigValues(config map[string]any, schema *Schema) error {
+	for k, v := range schema.Properties {
+		if _, ok := config[k]; !ok {
+			if v.Default == nil {
+				return fmt.Errorf("input parameter %s is not defined in config", k)
+			}
+			config[k] = v.Default
+		}
+	}
+	return nil
+}
+
+func validateConfigValueTypes(config map[string]any, schema *Schema) error {
+	// validate types defined in config
+	for k, v := range config {
+		fieldMetadata, ok := schema.Properties[k]
```

Comment:
This is called `fieldInfo` in an earlier function, let's use the same everywhere.

Reply:
done